pkg/cloudprovider/vsphereparavirtual/loadbalancer_test.go

```diff
@@ -44,385 +40,310 @@ import (
-		testK8sServiceNameSpace string
-
-		// FakeClientWrapper allows functions to be replaced for fault injection
-		fcw *util.FakeClientWrapper
-	)
-
-	newTestLoadBalancer := func() (cloudprovider.LoadBalancer, error) {
-		scheme := runtime.NewScheme()
-		_ = vmopv1alpha1.AddToScheme(scheme)
-		fc := fakeClient.NewFakeClientWithScheme(scheme)
-		fcw = util.NewFakeClientWrapper(fc)
-		vms := vmservice.NewVMService(fcw, testClusterNameSpace, &testOwnerReference)
-		return &loadBalancer{
-			vmService: vms,
-		}, nil
+func TestNewLoadBalancer(t *testing.T) {
+	testCases := []struct {
+		name    string
+		testEnv *envtest.Environment
+		err     error
+	}{
+		{
+			name:    "NewLoadBalancer: when everything is ok",
+			testEnv: &envtest.Environment{},
+			err:     nil,
+		},
 	}
 
-	BeforeEach(func() {
-		lb, _ = newTestLoadBalancer()
-		testK8sServiceName = "test-lb-service"
-		testK8sServiceNameSpace = "test-service-ns"
-	})
-
-	Describe("NewLoadBalancer", func() {
-		Context("when everything is ok", func() {
-			var (
-				err     error
-				cfg     *rest.Config
-				testEnv *envtest.Environment
-			)
-
-			It("should not return error", func() {
-				testEnv = &envtest.Environment{}
-				cfg, err = testEnv.Start()
-				Expect(err).ShouldNot(HaveOccurred())
-
-				_, err = NewLoadBalancer(testClusterNameSpace, cfg, &testOwnerReference)
-				if err != nil {
-					klog.Infof("nicole, error is ", err)
-				} else {
-					klog.Infof("nicole, lb error is nothing")
-				}
-				Expect(err).ShouldNot(HaveOccurred())
-
-				err = testEnv.Stop()
-				Expect(err).ShouldNot(HaveOccurred())
-			})
-		})
-	})
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			cfg, err := testCase.testEnv.Start()
+			assert.NoError(t, err)
 
-	Describe("GetLoadBalancer", func() {
-		var (
-			err    error
-			exists bool
-		)
+			_, err = NewLoadBalancer(testClusterNameSpace, cfg, &testOwnerReference)
+			assert.NoError(t, err)
+			assert.Equal(t, testCase.err, err)
 
-		JustBeforeEach(func() {
-			_, exists, err = lb.GetLoadBalancer(context.Background(), testClustername, testK8sService)
+			err = testCase.testEnv.Stop()
+			assert.NoError(t, err)
 		})
+	}
+}
 
-		Context("when VMService is created", func() {
-			BeforeEach(func() {
-				testK8sService = &v1.Service{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      testK8sServiceName,
-						Namespace: testK8sServiceNameSpace,
-					},
-				}
-				_, err := lb.EnsureLoadBalancer(context.Background(), testClustername, testK8sService, []*v1.Node{})
-				Expect(err).Should(Equal(vmservice.ErrVMServiceIPNotFound))
-			})
-			AfterEach(func() {
-				err := lb.EnsureLoadBalancerDeleted(context.Background(), testClustername, testK8sService)
-				Expect(err).ShouldNot(HaveOccurred())
-			})
-			It("should not return error", func() {
-				Expect(exists).Should(Equal(true))
-				Expect(err).ShouldNot(HaveOccurred())
-			})
-		})
+func TestGetLoadBalancer_VMServiceNotFound(t *testing.T) {
```

Comment:
These several `TestGetLoadBalancer` could be refactored to single func with list of testcases

Reply:
actually - the test flow is different
the first one only calls GetLB() but the second one also calls EnsureLB and EnsureLBDeleted functions
we can combine them but it may not be readable I think 🤔

Reply:
ok hm. Then let's keep it